File: 200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    // bool check(vector<vector<int>>&mat,int i,int j){
    //     if(i>0 && mat[i-1][j]==1) return false;
    //     if(i<n-1 && mat[i+1][j]==1) return false;
    //     if(j>0 && mat[i][j-1]==1) return false;
    //     if(j<n-1 && mat[i][j+1]==1) return false;
    // }
    void solve(vector<vector<char>>&grid,vector<vector<int>>&dp,int i,int j,int n,int m){
        if(i<0 || j<0 || i==n || j==m || grid[i][j]=='0' || dp[i][j]==1 ) return;
        dp[i][j]=1;
        solve(grid,dp,i-1,j,n,m);
        solve(grid,dp,i,j+1,n,m);
        solve(grid,dp,i+1,j,n,m);
        solve(grid,dp,i,j-1,n,m);
    }
    int numIslands(vector<vector<char>>& grid) {
        int n=grid.size(),m=grid[0].size();
        vector<vector<int>>dp(n,vector<int>(m,0));
        int cnt=0;
        // for(int i=0;i<n;i++){
        //     for(int j=0;j<m;j++){
        //         cout<<dp[i][j]<<" ";
        //     }
        //     cout<<endl;
        // }
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]=='1' && dp[i][j]==0){
                    cnt++;
                    solve(grid,dp,i,j,n,m);
                }
            }
        }
        return cnt;
    }
    // int numIslands(vector<vector<char>>& grid) {
    //     int n=grid.size(),m=grid[0].size(),cnt=0;
    //     vector<vector<int>>dp(n,vector<int>(m,0));
    //     for(int i=0;i<n;i++){
    //         for(int j=0;j<m;j++){
    //             if(grid[i][j]=='1'){
    //                 if(dp[i][j]==0){
    //                     cnt++;
    //                     cout<<i<<" "<<j<<endl;
    //                 }
    //                 if(i<n-1) dp[i+1][j]=1;
    //                 if(j<m-1) dp[i][j+1]=1;
    //                 dp[i][j]=1;
    //             }
    //         }
    //     }
    //     return cnt;
    // }
};

File: 200-number-of-islands/number-of-islands.java
class Solution {
    
    private char[][] grid;
    private int n , m, count = 0;
    private boolean[][] visited;
    void dfs(int i , int j){
        if(i >= n || j >= m || i < 0 || j < 0 || visited[i][j]==true || grid[i][j] == '0' ) return;
        visited[i][j]=true;
        dfs(i+1,j);
        dfs(i,j+1);
        dfs(i-1,j);
        dfs(i,j-1);
    }

    public int numIslands(char[][] map) {
        grid = map;
        n = grid.length;
        m = grid[0].length;
        visited= new boolean[n][m];
        for(int i = 0; i < n; i++){
            for(int j = 0 ; j < m ; j++){
                if(grid[i][j]=='1' && !visited[i][j]){
                    dfs(i,j);
                    count++;
                }
            }
        }
        return count;
    }
}